### PR TITLE
Prefer processes in coordinator selection for multi-region clusters

### PR DESCRIPTION
**Pull request by johscheuer**
# Description

Fixes: https://github.com/FoundationDB/fdb-kubernetes-operator/issues/2034

## Type of change

*Please select one of the options below.*

- New feature (non-breaking change which adds functionality)

## Discussion

Most of the code changes are adjustments for the testing code, to make sure we add some randomness.

## Testing

Added additional unit tests and modified them.

## Documentation

-

## Follow-up

-

### CODE REVIEW

File: internal/locality/locality.go [hunk 1]
```diff
@@ -180,6 +185,29 @@ func ChooseDistributedProcesses(cluster *fdbv1beta2.FoundationDBCluster, process
 		currentLimits[field] = 1
 	}
 
+	if constraint.SelectingCoordinators {
+		// If the cluster runs in a multi-region config and has more than 3 DCs we want to make sure that the primary is
+		// preferred for the 3 coordinators.
+		// See: https://github.com/FoundationDB/fdb-kubernetes-operator/issues/2034
+		if cluster.Spec.DatabaseConfiguration.UsableRegions > 1 && cluster.Spec.DatabaseConfiguration.CountUniqueDataCenters() > 3 {
```
Comment (johscheuer) [inline internal/locality/locality.go:192]: This works for some specific setups e.g. 4 different DCs but not for some more exotic setups like having 1 primary dc with 2 different primary satellites and having 1 remote and 2 different remote satellites (that would be 6 unique dcs).

**Review by johscheuer — COMMENTED** [review]
The new benchmark will present a result like this:


```text
------------------------------
• [0.035 seconds]
Localities chooseDistributedProcesses when a multi-region cluster is used with 4 dcs when measuring the performance for choose distributed processes [measurement]
/Users/jscheuermann/go/src/github.com/FoundationDB/fdb-kubernetes-operator/internal/locality/locality_test.go:630

  Report Entries >>
  Choose Distributed Processes - /Users/jscheuermann/go/src/github.com/FoundationDB/fdb-kubernetes-operator/internal/locality/locality_test.go:636 @ 06/04/24 11:52:09.548
    Choose Distributed Processes
    Name                                  | N  | Min   | Median | Mean  | StdDev | Max
    ==================================================================================
    ChooseDistributedProcesses [duration] | 50 | 300µs | 400µs  | 400µs | 100µs  | 1ms
```

File: internal/locality/locality_test.go [hunk 2]
```diff
@@ -138,105 +163,260 @@ var _ = Describe("Localities", func() {
...
+									Priority: 0,
+								},
+								{
+									ID:        remoteSatelliteID,
+									Priority:  1,
+									Satellite: 1,
+								},
+								{
+									ID:        primarySatelliteID,
+									Priority:  0,
```
Comment (nicmorales9) [inline internal/locality/locality_test.go:354]: why is primary lower than remote here? or does the priority not actually signify primary vs remote? (though I don't see a distinction made in the tests so I don't think it matters)

**Review by nicmorales9 — APPROVED** [review]
just a few things to double check but the core lgtm

File: api/v1beta2/foundationdb_database_configuration.go [hunk 3]
```diff
@@ -562,6 +563,58 @@ func (configuration DatabaseConfiguration) GetNextConfigurationChange(finalConfi
...
+// it will only be counted as a main dc.
+func (configuration DatabaseConfiguration) GetMainDCsAndSatellites() (map[string]None, map[string]None) {
+	mainDCs := map[string]None{}
+	satellites := map[string]None{}
+
+	for _, region := range configuration.Regions {
+		for _, dataCenter := range region.DataCenters {
+			if dataCenter.Satellite == 0 {
+				mainDCs[dataCenter.ID] = None{}
+				// If the dc ID is present as satellite we will remove it.
```
[the remark this answers is not in the document]
Comment (johscheuer) [reply api/v1beta2/foundationdb_database_configuration.go:599]: That could be the case in a triplet configuration, e.g. https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/config/tests/multi_dc/final.yaml#L25-L45

File: internal/locality/locality_test.go (hunk 2 above)
Comment (johscheuer) [reply internal/locality/locality_test.go:354]: This is the config for the remote side and the remote side should prefer the remote satellite to store mutations and not the primary satellite. See: https://github.com/apple/foundationdb/wiki/Multi-Region-Replication#concepts

**Pull request merged**